File: public/trainmodel.js
async function trainModel() {
    // Load preprocessed data from CSV
    const response = await fetch('model/preprocessed_data.csv');
    const csvData = await response.text();

    // Parse CSV data using PapaParse
    const parsedData = Papa.parse(csvData, {
        header: true,
    });

    // Filter out rows with missing or undefined messageContent or labels
    const filteredData = parsedData.data.filter(row => row.messageContent && row.label !== undefined);

    // Store the total number of SMS messages used for training
    const totalSMSCount = filteredData.length;

    // Save the total SMS count to local storage for later use
    localStorage.setItem('totalSMSCount', totalSMSCount);

    // Map the data to messages and labels
    const messages = filteredData.map(row => row.messageContent);
    const labels = filteredData.map(row => parseInt(row.label));

    // Create a word index
    const wordIndex = {};
    let index = 1;
    messages.forEach(msg => {
        msg.toLowerCase().split(' ').forEach(word => {
            if (!wordIndex[word]) {
                wordIndex[word] = index++;
            }
        });
    });

    console.log(wordIndex);
    // Save the word index to a JSON file
    await fetch('model/word_index.json', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(wordIndex),
    });

    // Convert messages to sequences of integers based on the word index
    const sequences = messages.map(msg =>
        msg.split(' ').map(word => wordIndex[word] || 0)
    );

    // Pad sequences to the same length
    const maxlen = 100;
    const paddedSequences = sequences.map(seq => {
        if (seq.length > maxlen) {
            return seq.slice(0, maxlen);
        } else {
            return [...seq, ...Array(maxlen - seq.length).fill(0)];
        }
    });

    // Convert data to tensors
    const inputTensor = tf.tensor2d(paddedSequences);
    const labelTensor = tf.tensor2d(labels, [labels.length, 1]);

    // Define the model
    const model = tf.sequential();
    model.add(tf.layers.embedding({ inputDim: Object.keys(wordIndex).length + 1, outputDim: 16, inputLength: maxlen }));
    model.add(tf.layers.globalAveragePooling1d());
    model.add(tf.layers.dense({ units: 16, activation: 'relu' }));
    model.add(tf.layers.dense({ units: 1, activation: 'sigmoid' }));

    // Compile the model
    model.compile({
        optimizer: 'adam',
        loss: 'binaryCrossentropy',
        metrics: ['accuracy'],
    });

    // Train the model
    const history = await model.fit(inputTensor, labelTensor, {
        epochs: 10,
        validationSplit: 0.2,
        batchSize: 32,
    });

    console.log('Model training complete.');

    // Save the model to local storage or IndexedDB
    await model.save('localstorage://phishing-detection-model');
}

// Call the training function
trainModel();

File: public/model/train_model.py
import pandas as pd
import tensorflow as tf
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from sklearn.model_selection import train_test_split
import numpy as np
import json
import tensorflowjs as tfjs

# Load the preprocessed data
data = pd.read_csv('preprocessed_data.csv')

# Debug: Display the first few rows of the data
print(data.head())

# Preprocess the data
texts = data['messageContent'].astype(str).tolist()
labels = data['label'].tolist()

# Tokenize the text data
tokenizer = Tokenizer(oov_token='<OOV>')
tokenizer.fit_on_texts(texts)
sequences = tokenizer.texts_to_sequences(texts)
word_index = tokenizer.word_index

# Save the word index for later use in the web app
with open('word_index.json', 'w') as f:
    json.dump(word_index, f)

# Pad sequences to ensure uniform input size
maxlen = 100  # This should be the same as in your web app
padded_sequences = pad_sequences(sequences, maxlen=maxlen, padding='post')

# Convert labels to numpy array
labels = np.array(labels)

# Split the data into training and validation sets
X_train, X_val, y_train, y_val = train_test_split(padded_sequences, labels, test_size=0.2, random_state=42)

# Build the model
model = tf.keras.Sequential([
    tf.keras.layers.Embedding(input_dim=len(word_index) + 1, output_dim=16, input_length=100),  # Added input_length
    tf.keras.layers.GlobalAveragePooling1D(),
    tf.keras.layers.Dense(16, activation='relu'),
    tf.keras.layers.Dense(1, activation='sigmoid')
])

# Compile the model
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Train the model
history = model.fit(X_train, y_train, epochs=10, validation_data=(X_val, y_val), batch_size=32)

# Save the model in TensorFlow.js format
tfjs.converters.save_keras_model(model, 'tfjs_model')

print("Model training complete and saved in TensorFlow.js format.")